core/src/main/scala/org/http4s/multipart/MultipartParser.scala

```diff
@@ -0,0 +1,166 @@
+          emit(-\/(head)) ++ receive1(next => handle(cr ++ next))
+        }
+        else {
+          emit(-\/(bv)) ++ receiveLine(None)
+        }
+      }
+
+      leading map handle getOrElse receive1(handle)
+    }
+
+    def receiveCollapsedLine(leading: Option[ByteVector]): Process1[ByteVector, (ByteVector, Option[ByteVector])] = {
+      receiveLine(leading) pipe process1.fold((ByteVector.empty, None: Option[ByteVector])) {
+        case ((acc, tail), -\/(left)) => (acc ++ left, tail)
+        case ((acc, _), \/-((term, tail))) => (acc ++ term, tail)
+      }
+    }
+
+    lazy val start: Writer1[Headers, ByteVector, ByteVector] = {
+      def beginPart(leading: Option[ByteVector]): Process1[ByteVector, Option[ByteVector]] = {
+        def isStartLine(line: ByteVector): Boolean =
+          line.startsWith(startLine) && isTransportPadding(line.drop(startLine.size))
+
+        def isEndLine(line: ByteVector): Boolean =
+          line.startsWith(endLine) && isTransportPadding(line.drop(endLine.size))
+
+        def isTransportPadding(bv: ByteVector): Boolean =
+          bv.toSeq.find(b => b != ' ' && b != '\t').isEmpty
+
+        receiveCollapsedLine(leading) flatMap {
+          case (line, tail) if isStartLine(line) =>
+            emit(tail)
+          case (line, tail) if isEndLine(line) =>
+            halt
+          case (line, _) =>
+            fail(InvalidMessageBodyFailure(s"Expected multipart boundary, got: $line"))
+            halt
+        }
+      }
+
+      def go(leading: Option[ByteVector]): Writer1[Headers, ByteVector, ByteVector] = {
+        beginPart(leading) flatMap { tail =>
+          val expected = CRLF ++ startLine
+          for {
+            headerPair <- header(tail, expected)
+            (headers, tail2) = headerPair
+            spacePair <- receiveCollapsedLine(tail2)
+            (chomp, tail3) = spacePair // eat the space between header and content
+            part <- emit(-\/(headers)) ++ body(tail3, expected).flatMap {
+              case (chunk, None) =>
+                emit(\/-(chunk))
+              case (chunk, some @ Some(remainder)) =>
+                emit(\/-(chunk)) ++ go(some)
+            }
+          } yield part
+        }
+      }
+
+      go(None)
+    }
+
+    def header(leading: Option[ByteVector], expected: ByteVector): Process1[ByteVector, (Headers, Option[ByteVector])] = {
+      def go(leading: Option[ByteVector], expected: ByteVector): Process1[ByteVector, (Headers, Option[ByteVector])] = {
+        receiveCollapsedLine(leading) flatMap {
+          case (bv, tail) => {
+            if (bv == expected) {
+              halt
+            } else {
+              val headerM = for {
+                line <- bv.decodeAscii.right.toOption
+                idx <- Some(line indexOf ':')
+                if idx >= 0
+                if idx < line.length - 1
+              } yield Header(line.substring(0, idx), line.substring(idx + 1).trim)
```

Comment:
These should only be headers that begin with `"Content-"`.  I don't see a sensible way to move this to the type level.